### ISSUE TITLE
sqlalchemy.exc.OperationalError: no such table: user

### ISSUE THREAD

**Issue by Tanker21** [miguelgrinberg/microblog#268]
I am working on chapter four, Databases, and am at the point of just adding the Post table to the database, and going into the python interpreter to play around creating users, but when I type in "db.session.commit()", I get this error:

Traceback (most recent call last):
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\engine\base.py", line 1276, in _execute_context
    self.dialect.do_execute(
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\engine\default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
sqlite3.OperationalError: no such table: user

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\orm\scoping.py", line 163, in do
    return getattr(self.registry(), name)(*args, **kwargs)
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\orm\session.py", line 1042, in commit
    self.transaction.commit()
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\orm\session.py", line 504, in commit
    self._prepare_impl()
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\orm\session.py", line 483, in _prepare_impl
    self.session.flush()
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\orm\session.py", line 2536, in flush
    self._flush(objects)
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\orm\session.py", line 2678, in _flush
    transaction.rollback(_capture_exception=True)
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\util\langhelpers.py", line 68, in __exit__
    compat.raise_(
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\util\compat.py", line 182, in raise_
    raise exception
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\orm\session.py", line 2638, in _flush
    flush_context.execute()
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\orm\unitofwork.py", line 422, in execute
    rec.execute(self)
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\orm\unitofwork.py", line 586, in execute
    persistence.save_obj(
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\orm\persistence.py", line 239, in save_obj
    _emit_insert_statements(
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\orm\persistence.py", line 1135, in _emit_insert_statements
    result = cached_connections[connection].execute(
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\engine\base.py", line 1011, in execute
    return meth(self, multiparams, params)
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\sql\elements.py", line 298, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\engine\base.py", line 1124, in _execute_clauseelement
    ret = self._execute_context(
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\engine\base.py", line 1316, in _execute_context
    self._handle_dbapi_exception(
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\engine\base.py", line 1510, in _handle_dbapi_exception
    util.raise_(
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\util\compat.py", line 182, in raise_
    raise exception
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\engine\base.py", line 1276, in _execute_context
    self.dialect.do_execute(
  File "D:\Documents\Learning\Python\Flask\microblog\venv\lib\site-packages\sqlalchemy\engine\default.py", line 593, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: user
[SQL: INSERT INTO user (username, email, password_hash) VALUES (?, ?, ?)]
[parameters: ('john', 'john@example.com', None)]
(Background on this error at: http://sqlalche.me/e/13/e3q8)

basically stating the table 'user' does not exist, even though I migrated and upgraded.  I got out of the virtual environment, closed all files, checked all code in my files to the author's code, restarted the virtual env, same error, so I deleted the entire migrations folder and walked through the steps again, but always get the same error, that user table does not seem to be getting created in the database.

**Comment by miguelgrinberg** [comment]
You need to describe what those steps that you followed are, because I'm sure you missed something in there. In particular it would be interesting to see the output of all the `flask db ...` commands that you issue.

**Comment by Tanker21** [comment]
I'm sure its something I missed, just can't determine what it is!

Everything seemed to work and follow what was in the walkthrough, up to the 'Play Time' section in Chapter 3.  I opened a Python interpreter window, within my venv of the microblog project, and typed in:
>>> from app import db
>>> from app.models import User, Post
>>> u = User(username='john', email='john@example.com')
>>> db.session.add(u)

and say no issues, but when I typed:
>>> db.session.commit()
 
I received the above error.  I did previously type in:

>>> from app.models import User
>>> u = User(username='susan', email='susan@example.com')
>>> u
<User susan>

so the models.py seems to be configured correctly, but something I'm doing in migrating the database is where the issue seems to be.

All I did for the db migration was (within the project venv):
(venv) PS D:\Documents\Learning\Python\Flask\microblog> flask db init
d:\documents\learning\python\flask\microblog\venv\lib\site-packages\flask_sqlalchemy\__init__.py:812: UserWarning: Neither SQLALCHEMY_DATABASE_URI nor SQLALCHEMY_BINDS is set. Defaulting SQLALCHEMY_DATABASE_URI to "sqlite:///:memory:".
  warnings.warn(
Creating directory D:\Documents\Learning\Python\Flask\microblog\migrations ...  done
Creating directory D:\Documents\Learning\Python\Flask\microblog\migrations\versions ...  done
Generating D:\Documents\Learning\Python\Flask\microblog\migrations\alembic.ini ...  done
Generating D:\Documents\Learning\Python\Flask\microblog\migrations\env.py ...  done
Generating D:\Documents\Learning\Python\Flask\microblog\migrations\README ...  done
Generating D:\Documents\Learning\Python\Flask\microblog\migrations\script.py.mako ...  done
Please edit configuration/connection/logging settings in 'D:\\Documents\\Learning\\Python\\Flask\\microblog\\migrations\\alembic.ini' before proceeding.
(venv) PS D:\Documents\Learning\Python\Flask\microblog>

I then type in: 

(venv) PS D:\Documents\Learning\Python\Flask\microblog> flask db migrate -m "users table"
d:\documents\learning\python\flask\microblog\venv\lib\site-packages\flask_sqlalchemy\__init__.py:812: UserWarning: Neither SQLALCHEMY_DATABASE_URI nor SQLALCHEMY_BINDS is set. Defaulting SQLALCHEMY_DATABASE_URI to "sqlite:///:memory:".
  warnings.warn(
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.autogenerate.compare] Detected added table 'user'
INFO  [alembic.autogenerate.compare] Detected added index 'ix_user_email' on '['email']'
INFO  [alembic.autogenerate.compare] Detected added index 'ix_user_username' on '['username']'
INFO  [alembic.autogenerate.compare] Detected added table 'post'
INFO  [alembic.autogenerate.compare] Detected added index 'ix_post_timestamp' on '['timestamp']'
Generating D:\Documents\Learning\Python\Flask\microblog\migrations\versions\e7a98db53d64_users_table.py ...  done
(venv) PS D:\Documents\Learning\Python\Flask\microblog>

(NOTE: this also has Post table in it, since I also modified the models.py file)

then ran:

(venv) PS D:\Documents\Learning\Python\Flask\microblog> flask db upgrade
d:\documents\learning\python\flask\microblog\venv\lib\site-packages\flask_sqlalchemy\__init__.py:812: UserWarning: Neither SQLALCHEMY_DATABASE_URI nor SQLALCHEMY_BINDS is set. Defaulting SQLALCHEMY_DATABASE_URI to "sqlite:///:memory:".
  warnings.warn(
INFO  [alembic.runtime.migration] Context impl SQLiteImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.runtime.migration] Running upgrade  -> e7a98db53d64, users table
(venv) PS D:\Documents\Learning\Python\Flask\microblog>

which seemed to update the database (I'm not seeing any errors reported).  I then continue through the walkthrough to adding the Post table in models.py, then run the migrate and upgrade again, with no errors reported, then get to the 'Play Time' section where I test this, and get the error that the table does not exist.  So, I'm not sure what I am missing or running incorrectly, everything seems to be correct and the same as the walkthrough code.

I pushed my code to my github (https://github.com/Tanker21/flask_microblog) if that would help any.

**Comment by miguelgrinberg** [comment]
This is the problem:

> UserWarning: Neither SQLALCHEMY_DATABASE_URI nor SQLALCHEMY_BINDS is set. Defaulting SQLALCHEMY_DATABASE_URI to "sqlite:///:memory:".

You have not configured your database correctly, the SQLALCHEMY_DATABASE_URI needs to be set with the URL that points to your database. Start from the config.py file to determine why this setting isn't set.

**Comment by Tanker21** [comment]
That was it, I had two letters transposed in 'SQLALCHEMY_DATABASE_URI'!  I fixed my typo, deleted the migrations folder, rebuilt the database starting from flask db init...and now can see my added users in the database!  Thank you for your time on this!

Just for clarity, you noticed the issue from those first lines of the error message stating "SQLALCHEMY_DATABASE_URI nor SQLALCHEMY_BINDS is set. Defaulting SQLALCHEMY_DATABASE_URI to "sqlite:///:memory:", correct?  I was thinking that was related to the OR statement in setting those configs (if environment var not found, use basedir path and app.db), but apparently that is not what was occurring?

Again, thanks for your time on this!

**Comment by miguelgrinberg** [comment]
Right, this is a message from the Flask-SQLAlchemy extension that indicates the extension does not have a configured database.

**Comment by Tanker21** [comment]
Roger that.  Thanks, I will close this question since I think I'm good now and can continue forward.
